server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogramTests.java

```diff
@@ -0,0 +1,353 @@
+    }
+
+    /**
+     * When buckets have the same min after the reduce phase, they should be merged.
+     */
+    public void testSameMinMerge() {
+        InternalVariableWidthHistogram dummy_histogram = createTestInstance();
+        List<InternalVariableWidthHistogram.Bucket> buckets = new ArrayList<>();
+        for (long value : new long[]{1, 100, 700}) {
+            InternalVariableWidthHistogram.Bucket.BucketBounds bounds;
+            if(value == 1 || value == 100) {
+                bounds = new InternalVariableWidthHistogram.Bucket.BucketBounds(
+                    1, value
+                );
+            } else{
+                bounds = new InternalVariableWidthHistogram.Bucket.BucketBounds(
+                    value, value + 1
+                );
+            }
+            InternalVariableWidthHistogram.Bucket bucket = new InternalVariableWidthHistogram.Bucket(
+                value, bounds, 1, format, InternalAggregations.EMPTY
+            );
+            buckets.add(bucket);
+        }
+        InternalVariableWidthHistogram histogram = dummy_histogram.create(buckets);
+
+        MockBigArrays bigArrays = new MockBigArrays(new MockPageCacheRecycler(Settings.EMPTY), new NoneCircuitBreakerService());
+        ScriptService mockScriptService = mockScriptService();
+
+        MultiBucketConsumerService.MultiBucketConsumer bucketConsumer = new MultiBucketConsumerService.MultiBucketConsumer(DEFAULT_MAX_BUCKETS,
+            new NoneCircuitBreakerService().getBreaker(CircuitBreaker.REQUEST));
+        InternalAggregation.ReduceContext context = InternalAggregation.ReduceContext.forFinalReduction(
+            bigArrays, mockScriptService, bucketConsumer, PipelineAggregator.PipelineTree.EMPTY);
+
+        ArrayList<InternalAggregation> aggs = new ArrayList<>();
+        aggs.add(histogram);
+        List<InternalVariableWidthHistogram.Bucket> reduced_buckets = ((InternalVariableWidthHistogram) histogram.reduce(aggs, context)).getBuckets();
+
+        // Expected clusters: [ (1), (100), (700) ]
+        // Expected clusters after same min merge: [ (1, 100), (700) ]
+        // Expected centroids: [ 101/2, 700 ]
+        // Expected keys: [ 1, 700 ]
+        // Expected doc counts: [2, 1]
+        double double_error = 1d / 10000d;
+        assertEquals(2, reduced_buckets.size());
+        assertEquals((101d/2d), reduced_buckets.get(0).centroid(), double_error);
+        assertEquals(1d, reduced_buckets.get(0).getKeyForXContent(), double_error);
+        assertEquals(2, reduced_buckets.get(0).getDocCount());
+        assertEquals(700d, reduced_buckets.get(1).centroid(), double_error);
+        assertEquals(700d, reduced_buckets.get(1).getKeyForXContent(), double_error);
+        assertEquals(1, reduced_buckets.get(1).getDocCount());
+    }
+
+    @Override
+    protected void assertReduced(InternalVariableWidthHistogram reduced, List<InternalVariableWidthHistogram> inputs) {
```

Comment:
Just naming the tests `testWhatever` will cause them to run. This method is about asserting that a bunch of random inputs reduce the way you expect them to. By default it calls `createTestInstance` a bunch of times to make data to reduce. In some cases that doesn't make and you want to produce a more realistic random array to reduce. In those cases you'd override `randomResultsToReduce`. But this method should still assert that `reduced` is "correct" for `inputs`.

The idea here is that you can be slow and exhaustive and "obvious" rather than on the reduce implementation in production code where you have to be more efficient. In your case, though, reduction is quite complex. So I'm not sure what the right balance is here. I know the `auto_date_histogram` is kind of funky here because it feels like it is just reproducing the production code. Which doesn't feel good. I'm honestly not sure what to do here. But I know that random tests *in general* are surprisingly helpful at finding edge cases you'd never guess at.

Reply:
After I updated `createTestInstance()`, the `InternalAggregationTestCase.testReduceRandom()` method also started running useful tests and exposed some bugs that I've now fixed. 

I'm definitely with you on randomized testing. However, I've left `assertReduced()` empty (this is called at the very end of `testReduceRandom()`) because I don't know if randomization can be used here to check the reduction.

I'm not quite sure how we can determine what the correct reduction is without actually running the algorithm. I'm relying on the manual tests I've written for this and I can't see a way to get around that at the moment.